### PR TITLE
feat(qwen3-vl): Qwen3-VL Dense end-to-end on standard LLM path

### PR DESCRIPTION
**Pull request by JamesBrianD**
## Summary

First multimodal model on the **standard LLM serving path** (Scheduler / TpWorker / ModelRunner). The whole vision tower + LLM run in one jit / one mesh / one KV pool — no per-modality stage scheduler, no `stage_config.yaml`. This mirrors upstream sglang's `MiMoV2 / Qwen3VL` design and validates the pipeline extensions that will be reused by MiMo-V2.5 (issue #250).

Scope of this PR:
- ✅ Qwen3-VL Dense (`Qwen/Qwen3-VL-8B-Instruct`)
- ✅ Image input only
- ✅ TP=4, page_size>=16
- ❌ Video / Audio / MoE deferred

### Commits

| # | Commit | Purpose |
|---|---|---|
| 1 | `fix(jax): use jnp.clip(min=) keyword` | Unblock startup on jax>=0.10 |
| 2 | `feat(qwen3): add post_residual_addition hook` | LLM decoder gets an optional deepstack injection point (backwards-compatible) |
| 3 | `feat(scheduler): plumb per-request mm_inputs` | Carries multimodal inputs through `ModelWorkerBatch` → `ForwardBatch` (pytree aux_data) for monolithic VLMs |
| 4 | `feat(tokenizer): mm processor dispatch` | TokenizerManager lazy-loads a per-arch mm processor and runs it on requests with image_data |
| 5 | `feat(qwen3-vl): Qwen3-VL Dense E2E` | Config / ViT (Conv3D + learned pos_embed + 2D RoPE + main + 3 deepstack mergers) / LLM extension (MRoPE + input_embeds bypass + deepstack injection) / Wrapper / CPU processor / 750-key weight mapping |

### Architecture (high-level)

```
HTTP /v1/chat/completions
  → TokenizerManager (mm_processor dispatch)
    → Qwen3VLProcessor.process()                       [new in commit 5]
      → HF AutoProcessor → pixel_values + image_grid_thw + mrope
  → Scheduler (req.mm_inputs already supported)
  → ScheduleBatch.get_model_worker_batch
    → _collect_mm_inputs(reqs) → list[dict]            [commit 3]
  → ModelWorkerBatch.mm_inputs                          [commit 3]
  → ForwardBatch.init_new (mm_inputs in aux_data)       [commit 3]
  → Qwen3VLForConditionalGeneration.__call__:           [commit 5]
    → ViT(pixel_values, grid_thw)                       [commit 5]
    → splice main_features → input_embeds
    → scatter deepstack to padded zero tensor
    → Qwen3VLLanguageModel(input_embeds, deepstack):    [commit 5]
        → MRoPE 3D positions
        → input_embeds bypass
        → post_residual_addition per decoder layer       [commit 2]
  → logits → next token
```

## Test plan

Verified end-to-end on TPU v6e-4 with `Qwen/Qwen3-VL-8B-Instruct`. Launch:

```bash
python3 -m sgl_jax.launch_server \
    --model-path /models/Qwen3-VL-8B-Instruct --trust-remote-code \
    --tp-size 4 --context-length 4096 --page-size 16 \
    --disable-precompile --disable-radix-cache
```

- [x] **Text-only regression** — `"What is 2+2?"` → `"2 + 2"` (sane LLM response, no regression on Qwen3 path)
- [x] **Color recognition (5/5)** — pure 448×448 images of red/green/blue/white/black returned `"Red"`, `"Green"`, `"Blue"`, `"white"`, `"Black"` respectively
- [x] **Shape recognition** — red circle on white background → `"circle"`
- [x] **In-image OCR** — text "HELLO" → `"HELLO"`
- [x] **Weight loading** — all 750 HF checkpoint keys mapped, `"Qwen3-VL weights loaded successfully!"`
- [x] **Deepstack injection** — confirmed via `post_residual_addition` flowing into decoder layers 0/1/2

## Known limitations / out of scope

- `--disable-radix-cache` is required for now: radix cache keys on raw `input_ids` and would incorrectly share cached KV across different images (same `<|image_pad|>` token ids but different vision embeddings). Proper fix is to include `mm_inputs.hash` in the cache key (separate PR).
- Video / audio / MoE variants deferred.
- Logit-level parity vs HF transformers not yet automated (behavioral tests pass; numerical comparison to land in a follow-up).

Pilot context: primatrix/projects#256 (validates the standard-path approach for MiMo-V2.5 #250).

### CODE REVIEW

File: python/sgl_jax/srt/managers/tokenizer_manager.py [hunk 1]
```diff
@@ -337,11 +347,76 @@ def _validate_input_ids_in_vocab(self, input_ids: list[int], vocab_size: int) ->
...
+
+            vision_cfg = getattr(self.model_config.hf_config, "vision_config", None)
+            spatial_merge = getattr(vision_cfg, "spatial_merge_size", 2) if vision_cfg else 2
+            self.mm_processor = Qwen3VLProcessor(
+                model_path=self.server_args.model_path,
+                spatial_merge_size=spatial_merge,
+                trust_remote_code=self.server_args.trust_remote_code,
+            )
+            logger.info("Loaded Qwen3VLProcessor for %s", self.server_args.model_path)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (JamesBrianD) [reply python/sgl_jax/srt/managers/tokenizer_manager.py:376]: Addressed in 5ae7bdd1 — switched to `logger.exception` so the original traceback is logged. Kept the broad `except` because the deferred imports (transformers `AutoProcessor`, custom code) can fail with a wide variety of error types and we never want a single bad VLM config to crash unrelated text-only requests; the traceback in the log is enough to diagnose.

File: python/sgl_jax/srt/models/qwen3_vl.py [hunk 2]
```diff
@@ -0,0 +1,1141 @@
...
+        """
+        input_embeds = self.model.embed_tokens(forward_batch.input_ids)
+        # Reshard to replicated for jnp.where(size=N) -> cumsum requirement.
+        repl = NamedSharding(self.mesh, P())
+        input_ids_repl = jax.sharding.reshard(forward_batch.input_ids, repl)
+        input_embeds_repl = jax.sharding.reshard(input_embeds, repl)
+
+        n_image_tokens = vision_main.shape[0]
+        placeholder_mask = input_ids_repl == self.image_token_id
+        indices = jnp.where(placeholder_mask, size=n_image_tokens)[0]
```
[the remark this answers is not in the document]
Comment (JamesBrianD) [reply python/sgl_jax/srt/models/qwen3_vl.py]: Obsolete after the children refactor in e5241ea5. `_splice_vision_features` and its `jnp.where(placeholder_mask, size=N)` pad-to-0 trick are gone. The wrapper now uses `forward_batch.placeholder_positions` (built deterministically host-side in `_collect_mm_tensors`, sink slot at `input_ids.shape[0] - 1` instead of 0). The original concern — silently corrupting index 0 when the image-token count disagrees — no longer applies: any size mismatch is caught at `_collect_mm_tensors` time (`real_positions.size < n_real_image_tokens` triggers an explicit clamp + only-real-positions write), not by a wildcard fallback inside jit. Resolving.

File: python/sgl_jax/srt/model_executor/forward_batch_info.py [hunk 3]
```diff
@@ -234,6 +241,7 @@ def tree_flatten(self):
             "spec_algorithm": self.spec_algorithm,
             "capture_hidden_mode": self.capture_hidden_mode,
             "deterministic": self.deterministic,
+            "mm_inputs": self.mm_inputs,
```
[the remark this answers is not in the document]
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/forward_batch_info.py]: Addressed in e5241ea5 (the children refactor). `mm_inputs` is gone from `aux_data`. ForwardBatch now carries the bucket-padded device tensors (`pixel_values`, `placeholder_positions`, `cu_seqlens`) as pytree children, and `aux_data` only holds hashable Python primitives — `image_grid_thw: tuple[tuple[int,int,int],...]` and `n_real_images: int`. `image_grid_thw` is kept on the host because the ViT iterates over (h, w) as Python ints to build pos_embed / RoPE (jit specializes per unique tuple), which is the bucket-aware version of option (b) you suggested. Resolving.

File: python/sgl_jax/srt/models/qwen3_vl.py [hunk 4]
```diff
@@ -0,0 +1,1141 @@
...
+            and forward_batch.mm_inputs is not None
+        ):
+            vision_main, vision_deepstack = self._encode_and_splice_mm(forward_batch)
+            if vision_main is not None:
+                input_embeds, input_deepstack_embeds = self._splice_vision_features(
+                    forward_batch, vision_main, vision_deepstack
+                )
+                # Stash on forward_batch so Qwen3VLLanguageModel's input_embeds
+                # bypass picks it up (same convention as Qwen2.5-VL).
+                forward_batch.input_embedding = input_embeds
```
[the remark this answers is not in the document]
Comment (JamesBrianD) [reply python/sgl_jax/srt/models/qwen3_vl.py]: Agreed this is a hidden footgun. Filed as follow-up — will refactor `Qwen3VLLanguageModel.__call__` to accept an explicit `input_embeds` parameter so `forward_batch` stays immutable. Not blocking this PR (today's only reader is the LLM call right below, and the surrounding stage-path code already uses the same mutate-then-read pattern), but worth cleaning up before more callers grow.

File: python/sgl_jax/srt/models/qwen3_vl.py [hunk 5]
```diff
@@ -0,0 +1,1141 @@
...
+            )
+            layers_kv_fused.append(kv_fused)
+            layers_callback_flag.extend(callback_flag)
+
+        if residual is not None:
+            hidden_states += residual
+        # Last-layer deepstack: matches upstream's "self.norm(.., post_residual=last)" path
+        last_ds = self._get_deepstack_slice(len(self.layers) - 1, input_deepstack_embeds)
+        if last_ds is not None:
+            hidden_states = hidden_states + last_ds
```
[the remark this answers is not in the document]
Comment (JamesBrianD) [reply python/sgl_jax/srt/models/qwen3_vl.py]: Good catch — confirmed the branch was unreachable (`deepstack_embed_to_decoder_layer = [0, 1, 2]` and we queried `len(self.layers) - 1 = 35`). Removed the dead block + misleading comment in 5ae7bdd1. (Note: upstream sglang has the same dead code with the same indexing, so our 'matches upstream' comment was double-wrong.)

**Comment by JamesBrianD** [comment]
## Control-plane review: 4 additional findings

Did a deeper pass focused on the request → ViT scatter control plane. Skipping items already covered by the PR body (radix-cache collision, video/audio deferred scope) and by claude-bot's earlier review. Four new findings below — one is a real crash on DP>1, the rest are footguns / perf.

### 🔴 P0 — `_collect_mm_tensors(self.reqs, ...)` AssertionError under `dp_size > 1`

`python/sgl_jax/srt/managers/schedule_batch.py:2166`

```python
mm_tensors = _collect_mm_tensors(self.reqs, input_ids_cpu)
```

`reqs` is a field of `ScheduleReqsInfo`, so reading `self.reqs` goes through `ScheduleBatch.__getattr__` (line 778-784), which **asserts `dp_size == 1`**:

```python
assert self.dp_size == 1, (
    f"ScheduleBatch.{name} flat access is a dp=1 spec-decode shim; "
    f"dp={self.dp_size}>1 must use reqs_info[dp_rank].{name} (#1053 P1-5)"
)
```

Any multimodal request under DP>1 will crash here before reaching the model. Even if the assertion were bypassed, the cross-DP iteration order must match how `_merge_input_and_positions` (line 1774) packs `input_ids_cpu` — otherwise `pixel_values` row order and `np.where(input_ids == image_token_id)` ascending positions would misalign across DP ranks, silently scattering image-A's features into image-B's positions.

**Suggested fix:** flatten in DP-rank order to match the packer:

```python
all_reqs = [r for dp in range(self.dp_size) for r in self.reqs_info[dp].reqs]
mm_tensors = _collect_mm_tensors(all_reqs, input_ids_cpu)
```

The per-DP-rank padding gaps in `input_ids_cpu` contain zeros, not `<image_pad>`, so `np.where` is unaffected — only the req traversal order matters.

### 🔴 P1 — `video_data` / `audio_data` silently dropped

`python/sgl_jax/srt/managers/tokenizer_manager.py:384`

```python
def _has_multimodal_data(self, obj) -> bool:
    image_data = getattr(obj, "image_data", None)
    if image_data is None:
        return False
    ...
```

`io_struct.GenerateReqInput` exposes both `video_data` and `audio_data` (lines 253-256), but `_has_multimodal_data` only checks `image_data`. A request that sets only `video_data` or `audio_data` (no image) falls through to the **text-only path** — the modality is silently dropped and the model receives unrelated text generation.

The PR body explicitly defers video/audio, which is fine, but the failure mode should be **loud, not silent**. Suggest raising `NotImplementedError` in `_has_multimodal_data` (or in the request validation layer) when `video_data` / `audio_data` is set on a `Qwen3VLProcessor`-served model.

### ⚠️ P2 — `_pick_bucket` fallback triggers jit recompile for large images

`python/sgl_jax/srt/managers/schedule_batch.py:2633`

```python
def _pick_bucket(value: int, buckets: tuple[int, ...]) -> int:
    for b in buckets:
        if b >= value:
            return b
    return value  # <-- returns value itself when larger than max bucket
```

With `_MM_PATCH_BUCKETS = (256, 1024, 4096)`, any image exceeding ~1024×1024 effective resolution (at `patch_size=16, spatial_merge=2`) overflows the last bucket → ViT input shape becomes request-dependent → every new size triggers a fresh jit compilation. A user uploading a 1080p+ photo would hit a multi-minute cold start instead of a single recompile.

A few options, each with tradeoffs worth discussing:
- add larger buckets (e.g. `16384`) — costs more memory per ViT call
- clamp at processor stage by reducing `max_pixels` — rejects high-res info
- explicit error for oversize input — pushes the concern to the client

Happy to defer this until usage patterns inform the right tradeoff, but the silent recompile path should at least be documented.

### ⚠️ P2 — `compute_mrope_positions` is O(N²) on long multi-image prompts

`python/sgl_jax/srt/multimodal/manager/mrope_utils.py:42-52`

```python
for _ in range(remain_images + remain_videos):
    ed_image = (
        input_tokens.index(image_token_id, st)
        if remain_images > 0 and image_token_id in input_tokens[st:]
        else len(input_tokens) + 1
    )
```

`input_tokens[st:]` allocates a new list (O(N)) and `in` scans it (O(N)) — done every iteration, for `N_images + N_videos` iterations. For a 30k-token prompt with 4 images, this is ~120k of unnecessary work on CPU before the request even reaches the scheduler.

**Suggested fix** (single pass, same semantics):

```python
try:
    ed_image = input_tokens.index(image_token_id, st) if remain_images > 0 else len(input_tokens) + 1
except ValueError:
    ed_image = len(input_tokens) + 1
```

Same change applies to `ed_video` on the next branch.

---

Happy to send PRs for any of these — A is the most urgent since it's a hard crash gating DP+VL.

**Comment by JamesBrianD** [comment]
@JamesBrianD all four findings addressed. Verified each fix before pushing.

| # | Fix | Commit | Verification |
|---|---|---|---|
| 🔴 P0 | `_collect_mm_tensors` dp>1 crash | [`540ce07`](https://github.com/sgl-project/sglang-jax/commit/540ce071) | Switched to `all_reqs` flattened from `self.reqs_info` (DP-rank order matches `_merge_input_and_positions`' packing of `input_ids_cpu`, so `np.where` scatter placement stays consistent). Reproduced original assertion under dp_size=2; end-to-end tested with `--tp-size 4 --dp-size 2` on v6e-4 — 4 concurrent multimodal requests returned 200 OK with coherent image descriptions, no `AssertionError` in server log. |
| 🔴 P1 | `video_data` / `audio_data` silently dropped | [`b0d574e`](https://github.com/sgl-project/sglang-jax/commit/b0d574eb) | Raises `NotImplementedError` in `_has_multimodal_data` using the existing `has_valid_data` helper, so empty lists / None still take the text-only path. Tested 7 boundary cases: text-only / image only / image empty / video only / audio only / video empty / image+video — all behave as expected. |
| ⚠️ P2 | `_pick_bucket` silent jit recompile | [`4cfcf85`](https://github.com/sgl-project/sglang-jax/commit/4cfcf85b) | Took the "at least documented" path — `logger.warning` on overflow, behavior unchanged. The larger-bucket / clamp / error tradeoff is deferred. Verified `_pick_bucket(500)` (in-bucket) and `_pick_bucket(4096)` (boundary) stay silent; `_pick_bucket(8000)` (overflow) logs the warning. Confirmed it also surfaces organically in serving (`value=4160 exceeds max bucket=4096` showed up when a 480×480 image hit the path). |
| ⚠️ P2 | `compute_mrope_positions` O(N²) | [`5971eeb`](https://github.com/sgl-project/sglang-jax/commit/5971eeb3) | Took the suggested `try/except` around `.index()` for both `ed_image` / `ed_video`. Verified bit-identical output against the original implementation on a 30k-token / 4-image stress input; ~1.23x faster end-to-end on that shape (the constant factor matters more as inputs grow). |

**Comment by JamesBrianD** [comment]
## MMMU val 256-sample regression check — passed

| Config | Wall time | mmmu_acc |
|---|---|---|
| Pre three-jit (commits e92182c..5971eeb, mem=0.75, default max_running) | 88 min | **0.5469** |
| **With three-jit split + cache fix** (mem=0.85, max_running=128) | **60 min** | **0.5469** |

Identical accuracy. ~1.47x throughput improvement, driven by the LLM JIT cache now being shared between text-only and multimodal extends (per `d38181e8` — see PR #1211 follow-up comment for the cache_miss diagnosis).

Same lmms-eval command both runs:
```
python3 -m lmms_eval --model openai_compatible \
  --model_args 'model=/models/Qwen3-VL-8B-Instruct,api_key=EMPTY,base_url=http://127.0.0.1:8100/v1/' \
  --tasks mmmu_val --batch_size 128 --limit 256 \
  --gen_kwargs 'max_new_tokens=4096'
```

Server: `--tp-size 4 --context-length 8192 --page-size 128 --disable-radix-cache --attention-backend fa` (no `--disable-precompile`, no `--multimodal` flag — both validated to auto-detect correctly).
